File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionRotateBy.java
package br.com.insanitech.spritekit;

/**
 * Created by anderson on 06/01/17.
 */

class SKActionRotateBy extends SKAction {
    private float startRadians;
    private float radians;

    public SKActionRotateBy(float radians) {
        this.radians = radians;
    }

    @Override
    void computeStart() {
        startRadians = getParent().zRotation;
    }

    @Override
    void computeAction(long elapsed) {
        switch (timingMode) {
            case EaseIn: {
                getParent().zRotation = SKEaseCalculations.easeIn(elapsed, startRadians, radians, duration);
            }
            break;

            case EaseOut: {
                getParent().zRotation = SKEaseCalculations.easeOut(elapsed, startRadians, radians, duration);
            }
            break;

            case EaseInEaseOut: {
                getParent().zRotation = SKEaseCalculations.easeInOut(elapsed, startRadians, radians, duration);
            }
            break;

            case Linear: {
                getParent().zRotation = SKEaseCalculations.linear(elapsed, startRadians, radians, duration);
            }
            break;
        }
    }

    @Override
    void computeFinish() {
        getParent().zRotation = startRadians + radians;
    }

    @Override
    boolean willHandleFinish() {
        return false;
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/actions/SKActionRotateBy.kt
package br.com.insanitech.spritekit.actions

import br.com.insanitech.spritekit.SKEaseCalculations
import br.com.insanitech.spritekit.SKNode

/**
 * Created by anderson on 06/01/17.
 */

internal class SKActionRotateBy(private val radians: Float) : SKAction() {
    private var startRadians: Float = 0.0f

    override fun computeStart(node: SKNode) {
        this.startRadians = node.zRotation
    }

    override fun computeAction(node: SKNode, elapsed: Long) {
        when (this.timingMode) {
            SKActionTimingMode.EaseIn -> {
                node.zRotation = SKEaseCalculations.easeIn(elapsed.toFloat(), this.startRadians, this.radians, this.duration.toFloat())
            }

            SKActionTimingMode.EaseOut -> {
                node.zRotation = SKEaseCalculations.easeOut(elapsed.toFloat(), this.startRadians, this.radians, this.duration.toFloat())
            }

            SKActionTimingMode.EaseInEaseOut -> {
                node.zRotation = SKEaseCalculations.easeInOut(elapsed.toFloat(), this.startRadians, this.radians, this.duration.toFloat())
            }

            SKActionTimingMode.Linear -> {
                node.zRotation = SKEaseCalculations.linear(elapsed.toFloat(), this.startRadians, this.radians, this.duration.toFloat())
            }
        }
    }

    override fun computeFinish(node: SKNode) {
        node.zRotation = this.startRadians + this.radians
    }
}
